pkg/services/ngalert/sender/sender.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	return s.sdManager.ApplyConfig(sdCfgs)
+}
+
+func (s *Sender) Run() {
+	s.wg.Add(2)
+
+	go func() {
+		if err := s.sdManager.Run(); err != nil {
+			s.logger.Error("failed to start the sender service discovery manager", "err", err)
+		}
+		s.wg.Done()
+	}()
+
+	go func() {
+		s.manager.Run(s.sdManager.SyncCh())
+		s.wg.Done()
+	}()
+}
+
+// SendAlerts sends a set of alerts to the configured Alertmanager(s).
+func (s *Sender) SendAlerts(alerts apimodels.PostableAlerts) {
+	if len(alerts.PostableAlerts) == 0 {
+		s.logger.Debug("no alerts to send to external Alertmanager(s)")
+		return
+	}
+	as := make([]*notifier.Alert, 0, len(alerts.PostableAlerts))
+	for _, a := range alerts.PostableAlerts {
+		na := alertToNotifierAlert(a)
+		as = append(as, na)
+	}
+
+	s.logger.Debug("sending alerts to the external Alertmanager(s)", "am_count", len(s.manager.Alertmanagers()), "alert_count", len(as))
+	s.manager.Send(as...)
+}
+
+// Stop shuts down the sender.
+func (s *Sender) Stop() {
+	s.sdCancel()
+	s.manager.Stop()
+	s.wg.Wait()
+}
+
+// Alertmanagers returns a list of the discovered Alertmanager(s).
+func (s *Sender) Alertmanagers() []*url.URL {
+	return s.manager.Alertmanagers()
+}
+
+// DroppedAlertmanagers returns a list of Alertmanager(s) we no longer send alerts to.
+func (s *Sender) DroppedAlertmanagers() []*url.URL {
+	return s.manager.DroppedAlertmanagers()
+}
+
+func buildNotifierConfig(cfg *ngmodels.AdminConfiguration) (*config.Config, error) {
+	amConfigs := make([]*config.AlertmanagerConfig, 0, len(cfg.Alertmanagers))
+	for _, amURL := range cfg.Alertmanagers {
+		u, err := url.Parse(amURL)
+		if err != nil {
+			return nil, err
+		}
+
+		sdConfig := discovery.Configs{
+			discovery.StaticConfig{
+				{
+					Targets: []model.LabelSet{{model.AddressLabel: model.LabelValue(u.Host)}},
+				},
+			},
+		}
+
+		amConfig := &config.AlertmanagerConfig{
+			APIVersion:              config.AlertmanagerAPIVersionV2,
+			Scheme:                  u.Scheme,
+			PathPrefix:              u.Path,
+			Timeout:                 model.Duration(defaultTimeout),
+			ServiceDiscoveryConfigs: sdConfig,
+		}
+
+		// Check the URL for basic authentication information first
+		if u.User != nil {
+			amConfig.HTTPClientConfig.BasicAuth = &common_config.BasicAuth{
+				Username: u.User.Username(),
+			}
+
+			if password, isSet := u.User.Password(); isSet {
+				amConfig.HTTPClientConfig.BasicAuth.Password = common_config.Secret(password)
+			}
+		}
+		amConfigs = append(amConfigs, amConfig)
+	}
+
+	notifierConfig := &config.Config{
+		AlertingConfig: config.AlertingConfig{
+			AlertmanagerConfigs: amConfigs,
+		},
+	}
+
+	return notifierConfig, nil
+}
+
+func alertToNotifierAlert(alert models.PostableAlert) *notifier.Alert {
+	ls := make(labels.Labels, 0, len(alert.Alert.Labels))
+	a := make(labels.Labels, 0, len(alert.Annotations))
+
+	for k, v := range alert.Alert.Labels {
+		ls = append(
+			ls,
+			labels.Label{Name: removeSpaces(k), Value: v},
+		)
```

Comment:
Nit: This could be a single line (even the below append), since this breakdown is not really helping much in line width (we have longer lines at other places).